backends/bmv2/base_test.py

```diff
@@ -982,6 +899,25 @@ def counter_dump_data(self, counter_name, direct=False):
                 counter_entries.append(entry)
         return counter_entries
 
+    def meter_write(self, meter_name, index, meter_config, direct):
+        req = self.get_new_write_request()
+        update = req.updates.add()
+        update.type = p4runtime_pb2.Update.MODIFY
+        entity = update.entity
+        if direct:
+            meter_write = entity.direct_meter_entry
+            meter_obj = self.get_obj("direct_meters", meter_name)
+            meter_write.table_entry.table_id = meter_obj.direct_table_id
```

Comment:
@antoninbas I have a question about direct_meter writes. I am trying to use the p4runtime API to craft an update message, but I am unable to get this  particular write to work. The message I am seeing is 
`* At index 0: INVALID_ARGUMENT, 'Missing non-ternary field in match key'`.

I am not exactly sure how to craft a match key here that makes sense. I tried various things but keep getting the same message. Have not found a similar example, other operations on direct externs seem to be reads.   Any advice?

Reply:
You are only setting the `table_id` field in `meter_write.table_entry`. This is not enough, every table entry has its own meter entry. You need to provide the full match key, like you would if you were doing a `MODIFY` on a `TableEntry`.

You need to be able to call `set_match_key` here, for the direct case: the index is replaced by a match key.

Reply:
I see, so the direct_meter match key needs to match exactly with an existing, already installed match key?

Reply:
For a write operation, yes.
The direct meter entry can also be configured at the time of inserting the table entry, to avoid 2 RPCs: https://github.com/p4lang/p4runtime/blob/90553b90a12ead5c19700e7fef21164dea5b6d22/proto/p4/v1/p4runtime.proto#L191

Reply:
Thanks! I got something to work. 
There is one edge case that is tricky. I am guessing the direct meter always requires an entry? For example, if you have this construct:
```
table meter_table {
    actions = {
        meter_assign;
    }
    key = {
        h.eth_hdr.dst_addr: exact;
    }
    size = 16384;
    default_action = meter_assign();
    meters = table_attached_meter;
}
```
where meter_assign is executed as default action, one might have no entry on the table. I am guessing this case is not supported? I tried abusing the API with `is_default_action` but it did not work.

Reply:
P4Runtime & PI support direct resources for the default table entry, but bmv2 does not.
When attempting this, an error should be returned and the simple_switch_grpc logs should read: `[PI] bmv2 does not support direct meters for default entries`.

Reply:
Thanks, that was quite helpful.